### PR TITLE
Simplify Vite config

### PR DESCRIPTION
**Pull request by Rich-Harris**
Another stepping stone on the way to #7967 (which at this point is less important than the codebase improvements it's driving).

Right now the `vite` directory is a bit of a mess. The code is full of indirection, and the logic is split up in ways that are quite hard to follow. This simplifies everything.

The `vite-plugin-sveltekit-compile` plugin should be renamed `vite-plugin-sveltekit-serve`, though that would break the Storybook integration, which removes that plugin from the array.

### CODE REVIEW

File: packages/kit/src/exports/vite/index.js [hunk 1]
```diff
@@ -333,17 +261,152 @@ function kit({ svelte_config }) {
...
+					__SVELTEKIT_ADAPTER_NAME__: JSON.stringify(kit.adapter?.name),
+					__SVELTEKIT_APP_VERSION_FILE__: JSON.stringify(`${kit.appDir}/version.json`),
+					__SVELTEKIT_APP_VERSION_POLL_INTERVAL__: JSON.stringify(kit.version.pollInterval),
+					__SVELTEKIT_DEV__: 'false',
+					__SVELTEKIT_EMBEDDED__: kit.embedded ? 'true' : 'false'
+				},
+				optimizeDeps: {
+					exclude: ['@sveltejs/kit']
+				},
+				publicDir: ssr ? false : kit.files.assets,
```
Comment (benmccann) [inline packages/kit/src/exports/vite/index.js:379]: this seems like something that belongs in the serve plugin and in fact it's being overridden there. does this one have any effect?

File: packages/kit/src/exports/vite/index.js [hunk 2]
```diff
@@ -240,16 +154,15 @@ function kit({ svelte_config }) {
...
 			const client_hooks = resolve_entry(svelte_config.kit.files.hooks.client);
-			if (client_hooks) {
-				allow.add(path.dirname(client_hooks));
-			}
+			if (client_hooks) allow.add(path.dirname(client_hooks));
 
 			// dev and preview config can be shared
 			/** @type {import('vite').UserConfig} */
-			const result = {
+			const new_config = {
```
Comment (benmccann) [inline packages/kit/src/exports/vite/index.js:165]: `new_config` sounds a bit like it's the complete config. maybe `additional_config` to make it clearer that Vite will combine it with the provided config?

File: packages/kit/src/exports/vite/index.js [hunk 3]
```diff
@@ -333,17 +261,152 @@ function kit({ svelte_config }) {
...
+							assetFileNames: `${prefix}/assets/[name]-[hash][extname]`,
+							hoistTransitiveImports: false
+						},
+						preserveEntrySignatures: 'strict'
+					},
+					target: ssr ? 'node16.14' : undefined,
+					// don't use the default name to avoid collisions with 'static/manifest.json'
+					manifest: 'vite-manifest.json'
+				},
+				define: {
```
Comment (benmccann) [inline packages/kit/src/exports/vite/index.js:369]: this kind of stuff seems like it belongs in the setup plugin

**Review by benmccann — COMMENTED** [review]
Can we combine the build and serve plugins back to just a compile plugin? I don't know that we gain a lot by splitting it. This change will likely break Storybook and Histoire because they probably would need to ignore the build plugin as well, so at least it'd be good to send them PRs with the new plugin names first. But I'm not sure it's worth doing that. The serving part is already well separated by living in `configureServer` and `configurePreviewServer`. It also seems like this PR has some config in the build plugin that I'd put in the serve plugin and vice versa. I think we could just ignore that complication by keeping it a single plugin.

File: packages/kit/src/exports/vite/index.js [hunk 4]
```diff
@@ -618,13 +620,64 @@ function kit({ svelte_config }) {
...
+		 */
+		async config(config) {
+			if (is_build) return;
+
+			// dev and preview config can be shared
+			/** @type {import('vite').UserConfig} */
+			const new_config = {
+				appType: 'custom',
+				base: svelte_config.kit.paths.base,
+				build: {
```
[the remark this answers is not in the document]
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:648]: No, this is serve configuration

File: packages/kit/src/exports/vite/index.js (hunk 1 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:379]: errr.... not sure. but presumably the `ssr ? false : kit.files.assets` thing is for a reason. if it's not having an effect then perhaps something is broken? need to investigate

File: packages/kit/src/exports/vite/index.js (hunk 3 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:369]: why? the `define` values for build are different than for dev

Comment (benmccann) [reply packages/kit/src/exports/vite/index.js:369]: I intended the setup plugin to contain all the setup regardless of whether it's dev or build - it can return different config depending on which command is run. Tools like Storybook will still need things like `__SVELTEKIT_DEV__` set appropriately in build mode.

File: packages/kit/src/exports/vite/index.js (hunk 1 above)
Comment (benmccann) [reply packages/kit/src/exports/vite/index.js:379]: Ah, this plugin only runs in build mode and the serve plugin only runs if not in build mode, so it won't override this. I'll resolve this comment

File: packages/kit/src/exports/vite/index.js (hunk 3 above)
Comment (Rich-Harris) [reply packages/kit/src/exports/vite/index.js:369]: Do they? I think that stuff's only use by SvelteKit internals that shouldn't be imported in a Storybook context. Ww _can_ put it in `setup` but it'll increase complexity

Comment (benmccann) [reply packages/kit/src/exports/vite/index.js:369]: Users may create components that reference [SvelteKit modules](https://kit.svelte.dev/docs/modules). If those modules use defines that are not available or set properly it will end up causing issues.

**Comment by Rich-Harris** [comment]
Reverted it to a single plugin and moved the `define` stuff around, even though the code is more confusing now.

I really dislike the concessions we're making here. Something is deeply wrong with how these tools interact with each other. I don't know what the correct solution is, but mutating an array of plugins to remove one that has a specific name is definitely not it.

File: packages/kit/test/apps/basics/test/client.test.js [hunk 5]
```diff
@@ -651,6 +651,21 @@ test.describe('Load', () => {
 			);
 		});
 	}
+
+	if (!process.env.DEV) {
+		test.skip('does not fetch __data.json if no server load function exists', async ({
```
Comment (benmccann) [inline packages/kit/test/apps/basics/test/client.test.js:656]: do you mean to always skip this test?

**Review by benmccann — APPROVED** [review]
We should probably at some point provide an API that lets you set things up without adding the compile plugin, so that other tools don't have to remove it. I think splitting the plugins in the way that we are probably makes sense and is necessary in order to support other Vite-based tools that want to utilize SvelteKit components, but I agree having to remove a plugin is a bit gross

File: packages/kit/test/apps/basics/test/client.test.js (hunk 5 above)
Comment (Rich-Harris) [reply packages/kit/test/apps/basics/test/client.test.js:656]: yes, we can't unskip it until we implement #7967. doing that _and_ reconfiguring the Vite plugin would have been a lot for one PR
